File: src/main/java/cz/vhromada/catalog/service/GenreService.java
package cz.vhromada.catalog.service;

import cz.vhromada.catalog.domain.Genre;
import cz.vhromada.catalog.repository.GenreRepository;
import cz.vhromada.common.service.AbstractMovableService;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cache.Cache;
import org.springframework.stereotype.Component;

/**
 * A class represents service for genres.
 *
 * @author Vladimir Hromada
 */
@Component("genreService")
public class GenreService extends AbstractMovableService<Genre> {

    /**
     * Creates a new instance of GenreService.
     *
     * @param genreRepository repository for genres
     * @param cache           cache
     * @throws IllegalArgumentException if repository genres is null
     *                                  or cache is null
     */
    @Autowired
    public GenreService(final GenreRepository genreRepository, @Value("#{cacheManager.getCache('catalogCache')}") final Cache cache) {
        super(genreRepository, cache, "genres");
    }

    @Override
    protected Genre getCopy(final Genre data) {
        final Genre newGenre = new Genre();
        newGenre.setName(data.getName());
        newGenre.setPosition(data.getPosition());

        return newGenre;
    }

}

File: src/main/kotlin/cz/vhromada/catalog/service/GenreService.kt
package cz.vhromada.catalog.service

import cz.vhromada.catalog.domain.Genre
import cz.vhromada.catalog.repository.GenreRepository
import cz.vhromada.common.entity.Account
import cz.vhromada.common.provider.AccountProvider
import cz.vhromada.common.provider.TimeProvider
import cz.vhromada.common.service.AbstractMovableService
import org.springframework.beans.factory.annotation.Value
import org.springframework.cache.Cache
import org.springframework.stereotype.Component

/**
 * A class represents service for genres.
 *
 * @author Vladimir Hromada
 */
@Component("genreService")
class GenreService(
        private val genreRepository: GenreRepository,
        accountProvider: AccountProvider,
        timeProvider: TimeProvider,
        @Value("#{cacheManager.getCache('catalogCache')}") cache: Cache) : AbstractMovableService<Genre>(genreRepository, accountProvider, timeProvider, cache, "genres") {

    override fun getAccountData(account: Account): List<Genre> {
        return genreRepository.findByAuditCreatedUser(account.id)
    }

    override fun getCopy(data: Genre): Genre {
        return data.copy(id = null)
    }

}
